### ISSUE TITLE
Improve powerSet performance

### ISSUE THREAD

**Issue by treeowl** [haskell/containers#890]
Obviously, there's only so much we can do, but we can do some. The most obvious optimization is to use a version of `insertMin` that takes a singleton tree as an argument instead of an element. But I can't help wondering if we can do better. The problem is obviously $\Omega(2^n)$, and our solution is $O(2^n\log n)$. Can we close that gap by either improving our solution or proving a tight(er) lower bound?

**Comment by treeowl** [comment]
@meooow25 , I started rambling about this in comments on your PR. Feel very welcome to think about it.

**Comment by treeowl** [comment]
We can certainly improve the *space* requirement for the result to $O(2^n)$, though I don't yet know whether or we can do so without making the time blow up. In particular, each set in the result with $k \ge 1$ elements can be built by `bin`ing two sets in the result of equal or nearly equal size with an element. So the trick is to get the right sets and the right element where we need them when we need them.

**Comment by treeowl** [comment]
Continuing to think out loud: imagine we build all the singletons first, then the 2-sets, then the 3-sets, etc. When we're working on the $k$-sets, we can hold on to the collections of $\lfloor k/2 \rfloor$-sets and $\lceil k/2 \rceil$-sets. The question is, how do we know which one to combine with which, and with which root element? I've drawn out the first few stages of this process, but it's hard to draw enough of it to look for patterns because the diagram grows exponentially! In case it helps anyone, here's a diagram for `powerSet [1..5]`. Each row represents a set size, and sets are lexicographically ordered from left to right (regardless of row).

```
[]
[1]                                                                                         [2]                                                           [3]       [4]   [5]
   [1,2]                                            [1,3]                    [1,4]   [1,5]      [2,3]                [2,4]       [2,5]    [3,4]     [3,5]     [4,5]
        [1,2,3]                    [1,2,4]  [1,2,5]        [1,3,4] [1,3,5]      [1,4,5]              [2,3,4] [2,3,5]       [2,4,5]            [3,4,5]
            [1,2,3,4] [1,2,3,5]        [1,2,4,5]             [1,3,4,5]                                 [2,3,4,5]
                 [1,2,3,4,5]
```

**Comment by treeowl** [comment]
For the next ... while ... I will assume that we're taking the power set of `[1..n]`. That can be generalized by tracking the position (in the original set) of the last element of a subset. Suppose we're building from columns `j` and `k` (which will either be identical or adjacent), so that each element we build will look like `p ++ [e] ++ q`, where `p` is from `j` and `q` is from `k`. We walk `p` through `j` from its beginning up until we reach a point where the largest element of `p` is too big to be able to choose `e` and still have enough room to take `q`. For each `p`, we need to be able to find the matching `q`s, and we have to do it *efficiently*. I don't *think* we can afford to walk `k` from the start for each `p`. So let's walk `j` once to break it up into lists by first element, and put those lists in an array indexed by first element. Now we can quickly find where the `q`s start for any given `p`.

The hardest remaining part seems to be fitting the collections of elements of each size together into a single `Set`. There's that wobbling back and forth in size. We have to know when to zig and when to zag.

**Comment by treeowl** [comment]
@ekmett Do you have any ideas, either in the direction I'm pointing or another one?

**Comment by meooow25** [comment]
I don't fully understand your last comment, but I can follow the ones before it. The diagram is also interesting.

Here has been my line of thought:

The current algorithm isn't useful in getting to $O(2^n)$ because of `insertMin`. If that were $O(1)$, we would be good. So this works in optimal time for lists, for instance, but not sets.

To construct sets in $O(1)$, we need left and right subtrees that are balanced wrt each other and the root element.
We can have an array `:: Array Int (Set a)` where the index is a bitmask denoting indices of the set of elements in the original set. Then we can build the sets for a mask by finding the middle set bit, taking that as root, then indexing into the array with the lower and upper parts of the mask for left and right subtrees.

If we find the middle bit by looping over the positions, we get $O(n 2^n)$ time but $O(2^n)$ memory which is nice.
If we instead binary search for the middle bit, given we have something like $O(1)$ `popCount`, we are back to $O(2^n \log n)$ time but still with $O(2^n)$ memory.
I don't know of a way to find the middle bit in $O(1)$, if we could do that we would be golden.

**Comment by meooow25** [comment]
It's actually not difficult.

```hs
middleBit 1 = 0
middleBit n
  | even n    = middleBit (n `div` 2) + 1
  | otherwise = middleBit (n - 1)
```

We can memoize for $O(1)$.

But I overlooked the fact that we need to convert from the mask order to the sorted order of sets. Trying to figure that out.

**Comment by treeowl** [comment]
I don't understand what you mean about a middle bit. For each suitable left child, we choose one or more suitable roots. For each such root, we choose one or more right children. For instance, suppose we're walking the collection of 2-sets and looking at the collection of 3-sets to form the collection of 6-sets (we could actually do it either way, depending on whether we want our trees to be heavier on the left or on the right). If we've walked up to `[1,3]` in the collection of 2-sets, then we need to consider, in turn, each root `4, 5, ...`. We can't use too big a root, because we need to have at least three greater elements in the right child, so we can stop considering roots once they get too big. Suppose we are at root 5, so we have a left subtree of `[1,3]` and a root of 5. We now need to use, as right children, all the 3-trees whose first element is at least 6. My array concept was intended to make it cheap to jump to those 3-trees, though I'm not sure if it's theoretically necessary.

**Comment by treeowl** [comment]
Ah, the pattern is obvious now that I see it. Silly me! We *always* go *up* by one in length after a subset that does not include the maximum value. We *always* go *down* by one in length otherwise. So once we have all the subsets, getting them in order is actually easy. We can make them a set using `fromDistinctAscList` (or a version thereof that knows the size in advance). Is there a better way to form the set in the end? Maybe, but let's get to something that works first and think about constant factors (and GC effects) later.

**Comment by meooow25** [comment]
Thanks for the explanation, I can see how you want to construct the sets, but I don't understand how it all ties together. Perhaps some code/pseudocode can make things clearer.

> I don't understand what you mean about a middle bit.

The middle set bit of `0b11011001` is bit 4, for example (0-indexed). To construct the set for this mask, we take element 4 as the root, the set we constructed for `0b1001` as the left subtree and that for `0b11000000` as the right subtree.

Here is my idea implemented. Not benchmarked or optimized.

```hs
import Data.Bits (bit, xor, (.&.))
import qualified Data.Array.Unboxed as UA
import qualified Data.List as L
import Data.Set.Internal

-- O(2^n)
powerSet :: forall a. Set a -> Set (Set a)
powerSet xs0 = buildSet 0 (bit n - 1)
  where
    n = size xs0
    xs = UA.listArray (0, n - 1) (toList xs0) :: UA.Array Int a

    -- Use unboxed array for better constant factor
    middleBit :: UA.Array Int Int
    middleBit = UA.listArray (1, bit n - 1) [f msk | msk <- [1 .. bit n - 1]]
      where
        f 1 = 0
        f msk | even msk  = (middleBit UA.! (msk `quot` 2)) + 1
              | otherwise = middleBit UA.! (msk - 1)

    sets :: UA.Array Int (Set a)
    sets = UA.listArray (0, bit n - 1) [f msk | msk <- [0 .. bit n - 1]]
      where
        f 0 = Tip
        f msk = bin x (sets UA.! mskl) (sets UA.! mskr)
          where
            m = middleBit UA.! msk
            x = xs UA.! m
            mskl = msk .&. (bit m - 1)
            mskr = msk `xor` bit m `xor` mskl

    -- Populate the array directly for better constant factor
    sortedMsks :: UA.UArray Int Int
    sortedMsks = UA.listArray (0, bit n - 1) (0 : L.foldl' step [] [n-1, n-2 .. 0])
      where
        step msks i = bit i : L.map (bit i +) msks ++ msks

    buildSet :: Int -> Int -> Set (Set a)
    buildSet l u
      | l > u     = Tip
      | otherwise = bin (sets UA.! msk) (buildSet l (m-1)) (buildSet (m+1) u)
      where
        m = l + (u - l) `quot` 2
        msk = sortedMsks UA.! m
```

**Comment by meooow25** [comment]
I agree that there's not much utility to this, personally I've just been treating this as a puzzle 😁 

> We really can't apply `powerSet` to anything substantially larger, because then it wouldn't fit in memory?

Probably so in practice. The formal limit for the input set size can be considered to be word_size - 1 since the output size must not exceed `maxBound :: Int`.

**Comment by treeowl** [comment]
@jwaldmann This is absolutely just a puzzle. IIRC, I added `powerSet` to `Data.Set` because @ekmett thought it should be there, and offered the simple implementation we currently use. Something `Set -> [Set]` would certainly be accepted if there's a use for it and a good implementation. Since it's not a "fundamental operation on mathematical sets", I don't think it can slip in under the door like `powerSet`.

Going back to `powerSet`: Dropping from $O(2^n \log n)$ space to $O(2^n)$ space (with optimal sharing) should buy usability for a few extra input elements. On the time side, the logarithmic factor probably doesn't matter, so if someone has an $O(2^n)$ space/$O(2^n \log n)$ time algorithm that actually runs faster than whatever $O(2^n)$ time algorithm we find, we can use the "theoretically slower" one instead.

**Comment by treeowl** [comment]
@meooow25 Your implementation looks interesting, but I can't make head or tail of how it works. Could you explain? How does it achieve lexicographical order? I also noticed a type error: unboxed arrays can't hold `Set`s.

**Comment by treeowl** [comment]
@meooow25 Your implementation doesn't pass the test suite, unfortunately.

**Comment by treeowl** [comment]
To clarify, it doesn't pass the test suite once the test has been changed as in #892.

**Comment by meooow25** [comment]
> How does that typecheck?

I didn't actually use unboxed arrays here (`Array` vs `UArray`). Perhaps the comment was confusing, I mean that as `TODO: Use unboxed array...`

> the subsets you produce are totally unbalanced—just lists

Oops, the middle bit logic was at fault. Updated, but it's a little more complex now.

> How does it achieve lexicographical order?

That's the `sortedMasks` part, which is the same as the current algorithm, only it constructs a list of bitmasks instead of a set of sets. Is there any other part you'd like me to explain?

**Comment by treeowl** [comment]
Yes, the whole thing? I don't understand the shape of it. Could you maybe open a PR with lots of comments? I'm a comment maximalist.

**Comment by meooow25** [comment]
I don't want to open a PR without actually testing that it is an improvement.

I'll try to explain with an example:
Consider that the input set is `{0,1,2}`.
Then we make an array `sets` to hold the power set, indexed by bitmasks `0b000` to `0b111`. The bits in a mask indicate what elements are in the set.
`sets = array (0b000,0b111) [(0b000,{}), (0b001,{0}), (0b010,{1}), (0b011,{0,1}), ... (0b111, {0,1,2})]`
To build these sets in constant time for any mask, we find the middle set bit, take the corresponding element as root, and pick the left and right subtrees out of `sets` by indexing with the masks we want (as I described in the comment before).
For example,
`sets ! 0b111 = bin 1 (sets ! 0b001) (sets ! 0b100)`
`sets ! 0b101 = bin 2 (sets ! 0b001) (sets ! 0b000)`
Now that we have all the sets, we need to convert them to lexicographical order. We use the current `powerSet` algorithm, which does build sets in the right order, to get the right order of bitmasks `[0b000, 0b001, 0b011, 0b111, 0b101 ... 0b100]`. Then we can pick the corresponding sets out of `sets` to get `[{}, {1}, {1,2}, {1,2,3}, {1,3}, ... {3}]`, and build them into the output set with `fromDistinctAscList`.

**Comment by treeowl** [comment]
@meooow25 FWIW, I would prefer to get a PR, even before perf testing. That'll sic CI on it, and make it easier for me (or others) to play with the implementation. I'm going to try to write an implementation based on my own idea to see how that compares, but I won't delay merging yours to do so. I suspect a lot of performance issues will come down to how badly we trash the generational hypothesis, which is mostly about how much long-lived intermediate structure we build. As I said, I don't understand your way yet. Mine (to the extend I've elaborated it so far) does not seem to have a good answer for that.

**Comment by ekmett** [comment]
> It's fascinating to follow your discussion, but, forgive me, where's the use case?

The primary usecase I have (and what was the original motivation behind the instance) is that this algorithm is useful when constructing NFAs from DFAs. I have code for that that exists, even if its somewhat slow and off hackage buried in the ekmett/coda repo.

In my particular case I wound up using a Lazy version of the Data.Set library to achieve that, but it also led to the original code for computing sums and products of sets.

**Comment by meooow25** [comment]
So I benchmarked my implementation, not very thoroughly, and it does worse than the current algorithm. Which is not too surprising.

> As far as I understand, the only intermediate thing should be this "array of sets, indexed by bitmasks"?

I believe the `middleBits` array would also stick around as long as `sets`, since the sets in `sets` depend on it.

> There must be something in Hacker's Delight, HAKMEM, or https://graphics.stanford.edu/~seander/bithacks.html

Thanks for the pointers, I just looked them up but they seem to have nothing about this. I did find a relevant post by searching: https://www.keithschwarz.com/binary-subsets/, but their algorithm is no better than ours.

I will hack on my algorithm's constant factor later and make a PR if it's promising.
I'm also interested to see @treeowl's approach realized.

**Comment by treeowl** [comment]
@meooow25 That's disappointing; I'd have expected an automatic performance boost from just reducing the result size. Two things to consider:

1. Make `sortedMasks` a right fold instead of a left fold. I'm not sure that'll help, but it probably won't hurt.
2. More significantly, don't use `DistinctAscList`. You're probably better off leaving out the empty set (to add later) and writing a custom function to build a full tree of a known size.

**Comment by treeowl** [comment]
One more question: aside from `gauge` benchmarks, did you try informal ones that calculate just one power set? That should let you use bigger inputs.

**Comment by treeowl** [comment]
> > a custom function to build a full tree of a known size.
> 
> this sounds useful in general? (`fromDistinctAscListN`, like https://hackage.haskell.org/package/vector-0.13.0.0/docs/Data-Vector.html#v:fromListN)

I agree. How should that work? Make a canonical skew binary representation and then `merge` the digits?

**Comment by treeowl** [comment]
@jwaldmann Good point. While the algorithm Okasaki gives is for the wrong order, the one we need is much simpler.

**Comment by treeowl** [comment]
@meooow25 I've been working on this yesterday and today. Aside from the parts that are tricky, there are also some subtleties I'm trying to work through regarding how to arrange things for best performance. One of the subtler ones is the order in which the final set should be constructed, which affects what arrays get constructed and freed when. Another tricky bit that may factor in: since the power set is (presumably) to be used as a `Set`, it's desirable for the subsets to get smaller the closer they are to the root of the power set. That way, doing something like deleting an element from the power set is as cheap as possible. Can that desire be reconciled with whatever is the most efficient construction order? Etc. I'm going to attempt to do whatever seems simplest for the moment and then revisit it. There are also a *lot* of fiddly bits for my limited brain to keep track of. If I don't get it all figured out tomorrow I'll put up the code I have so far with explanations and see if maybe you can take a crack at finishing it.

----

Regarding your version, I suspect the following issues are behind (at least some of) your performance problems:

1. The `msks` you use in `sortedMasks` is going to end up being shared by its two use sites, which might sound good but is actually very bad—something like half this list will end up actually realized in memory. It would be much better to find a different way to calculate `sortedMasks`; is there some pattern you can exploit from a different direction? I suspect that fixing this could make your solution competitive with (and quite likely better than) the current version.
2. `sets` is an utterly gargantuan array, which means that simply constructing it is quite expensive. The RTS will first allocate the array and fill it with `undefined`s; slow as that is, I think it's fine, and there's unlikely to be a good solution that *doesn't* pay that cost. The next step is another story: your program will allocate a thunk representing each subset and store it in the array. As usual, allocating an enormous number of long-lived objects (even rather small ones) is hard on the GC, so even this rather boring `listArray` will likely take quite a while for a "large" set. Once it's allocated, you'll walk the `sortedMasks` and thunks will get forced and so on. A (smaller?) problem happens on the other end; none of the array structure itself can be freed until you're all done.

**Comment by treeowl** [comment]
OK. So ... here's the current general status of my thingy. I'll go roughly from the top down.


### The final merge

Suppose we have a list of length $n+1$ of lists, where the $k$&ZeroWidthSpace;th list (starting at 0) contains the subsets of size $k$, in order. So for the set ${1,2,3}$, this list would be

```haskell
biglists123 =
[ [[]]
, [[1], [2], [3]]
, [[1,2], [1,3], [2,3]]
, [[1,2,3]] ]
```

How do we merge them in order? Define the following:

```haskell
data SmallerBigger
  = Smaller
 | Bigger
  deriving Show

enddir :: Int -> [SmallerBigger]
enddir n0 = Bigger : ends' n0 []
  where
    ends' 1 = (Smaller :)
    ends' n = (Bigger :) . ends' (n - 1) . ends' (n - 1)
```

Open up a zipper into the lists above. `enddir n` will give us a list showing which direction to take to get from each subset to the next one (it will have an extra `Smaller` at the end that we don't need).

```haskell
enddir 3 = [Bigger,Bigger,Bigger,Smaller,Smaller,Bigger,Smaller,Smaller]
```

So we can just move the zipper back and forth to grab the elements in order. Note that if it proves advantageous, we can instead work backwards, using

```haskell
enddirRev :: Int -> [SmallerBigger]
enddirRev n0 = List.drop 1 $ ends' n0 [Smaller]
  where
    ends' 1 = (Bigger :)
    ends' n = ends' (n - 1) . ends' (n - 1) . (Smaller :)
```

That will tell us which way to move as we work through *decreasing* lists in *reverse* lexicographical order.

### Making the lists

I'll start with an approximation and refine it some. Represent each list as a sequence of arrays of subsets (that sequence type could be `[]`, or `Array Int`, or `Data.Sequence.Seq`; I suspect `Array Int` will be best, but I'm not sure). Each inner array represents a subset of the appropriate size and a particular starting element.

Suppose we have the sequence of arrays of the subsets of size $\lfloor (k - 1)/2 \rfloor$ and a similar one of the ones of size $\lceil (k-1)/2 \rceil$. We need to get left subtrees from one, and right subtrees from the other; I'm not sure which should be which; for big- $O$, either will work. We'll use these to construct the subsets of size $k$. For each starting value (i.e., each element of the sequence of arrays of left subtrees), we produce a list of subsets by choosing each of the possible roots, and for each root choosing each of the possible starting values of right subtrees. This need to limit by starting values is the main reason why we have multiple arrays for each subset size. OK! So ... how do we know which roots are acceptable for a given left subset? We need to know which element (well, position in the underlying set) is the last one in that subset. We *could* store those separately, but that's inefficient. The solution is very much like `enddir` above:

```haskell
zoop :: Int -> [[[Int]]]
zoop n = List.map (List.map ($ [])) . zap . List.map (:) $ [1..n]
  where
    zap :: [[a] -> [a]] -> [[[a] -> [a]]]
    zap [] = []
    zap s@(_ : xs) = s : zap (List.map (List.foldr (.) id) (List.init $ List.tails xs))
```

For each subset size and each starting position, this produces a list of the first permissible root positions. That is, the position after the last position in each left subset of that size and starting position. We convert that list to an array for compact storage.

Once we have all the arrays, we can convert them to lists, concatenate them, and merge as shown above. As I said, however, this is an approximation. We only actually need to *store* the subsets of sizes up to $\lfloor n/2\rfloor$. So there's no benefit in actually building arrays for subsets above that size, and doing so is not cheap. So ... let's not. Once we've built all the arrays we need, we can leave the rest as lists, which will be consumed as they are produced.

**Comment by treeowl** [comment]
@meooow25 and @jwaldmann I'm feeling a bit stuck finishing my implementation and probably won't even be able to attempt it this week. Does either of you have the time/energy/brain space to take a shot at it? I'm happy to try to clear up any confusion in my description.

**Comment by meooow25** [comment]
I've been a bit busy lately. I can get back to this at the end of the week, but unlikely before that.

**Comment by meooow25** [comment]
I've been a bit busy lately. I can get back to this at the end of the week, but unlikely before that.

**Comment by treeowl** [comment]
I have absolutely no clue what your code is doing. Why calculate sums over the results? Could you test just calculating the results? And do it for inputs of size 20, 21, 22? The other test that would be interesting (I think) is to see how long it takes to delete each element from the power set; that tells us whether the set is arranged nicely.

**Comment by treeowl** [comment]
-

**Comment by treeowl** [comment]
It seems likely to me that we want to put the shortest subsets as close as possible to the root. I believe we can do that while maintaining strict (or at least nearly strict) balance. For example, half the subsets start with the first element, so I think the singleton of the second element should probably go at the root—it's $O(\log m)$ to compare that to a set of size $m$. It looks like that sort of trick should work recursively, getting progressively less effective going down and left.

**Comment by treeowl** [comment]
However, I'm more focused on construction time right now.

**Comment by meooow25** [comment]
@treeowl:
I understand your idea at a high-level, but I'm not able to follow the code you posted above.

Is `enddir` supposed to give the comparison of adjacent sets in the flattened list of sets? This doesn't seem to be so because it says `[1,3]` is smaller than `[1,2]`. And even if we had that, I don't understand how it helps in merging.

I also don't understand the output of `zoop`. `zoop 5 = [[[1],[2],[3],[4],[5]],[[2,3,4,5],[3,4,5],[4,5],[5]],[[3,4,5,4,5,5],[4,5,5],[5]],[[4,5,5,5],[5]],[[5]]]`. What does this mean?

---

I played a bit with my solution and tried to optimize it. Your suggestions were helpful, especially about avoiding thunks in `sets`.
Here's what I got it down to: [link](https://gist.github.com/meooow25/361ed5c5534ccbbb8fdac8d450e9a69d), and I'm curious if it can be optimized further.

And although the decrease is nice and it has been fun to mess with, the code is _waaay_ too complicated for little benefit, and I don't think it would be a good idea to replace the current code with this.

**Comment by treeowl** [comment]
@meooow25 The poorly named `enddir` indicates whether the *size* of each subset is one greater or one less than the size of the last one. So we'd see `Bigger` going from `[1]` to `[1,2]` and `Smaller` going from `[1,2,3]` to `[1,3]`.

`zoop 7 !! 3 !! 2` relates to subsets of size 3 that begin with element number 2 (counting from 0) of an underlying set of size 7. Each element is the number of the first element that could be used as the root given that left subtree. This zoop list is meant to be zipped with the corresponding array of subsets. I don't want to calculate a real example right now, but if you had subsets of `[0..20]` of size 3 that (counterfactually) looked like

```haskell
[0,2,9]
[2,5,7]
[3,4,5]
```

then the corresponding zoop values would be 10, 8, and 6. So for the first subset shown, the possible root positions would be `[10]` (anything bigger will run out of values for the right subtree). For the second, possible root values would be `[8..10]`, and for the third, possible root values would be `[6..10]`.

Question: is there an easy way to get each zoop value to be something more useful/meaningful than a number? Conceptually, it should be a list of `(root, [Set a])` pairs, where each root is paired with all the subsets of the appropriate size that can be its right children.

----

Note: my intuition keeps suggesting that we should build the power set in reverse, but I don't have proper evidence for that intuition.

**Comment by treeowl** [comment]
@AlekJSmith , thank you for looking into this. I haven't looked at your proposed pattern in detail, but I'm wondering how you propose to make use of it. Also, would you like to meet up in person to pair program this?

**Comment by AlekJSmith** [comment]
@treeowl meeting up in person sounds great to me, would Clydes work for you?

 My aim is given a set of size n map over the index of [1..(2^n)] and generate the entries of the powerset by:

1. directly calculating it given the index e.g. for the size 1 entries all you need to know is if the index is included in pattern 2; or 
2. composing the balanced binary tree for that set from the smaller sets elsewhere in the powerset index. Pattern 1 implies there should be a pattern for how larger (numeric) sets can be composed from smaller ones with just the powerset indices and I believe that larger (balanced binary tree) sets also have patterns for how they can be composed from smaller ones (Pattern 3 below) which can be determined from the set size which can be directly calculated from the powerset index from Pattern 1. 


**Possible Pattern 3 : the right branch of balanced binary trees only changes at given set sizes (as a consequence of rotation)**

Until the set reaches a certain size the right most branch of the tree stays the same, even when reaching the next set size for the right branch to change it is only to drop a level with a new root which is double the old one  e.g.

`putStrLn $ showTreeWith True False $ fromList [1..6]

4
+--2
|  +--1
|  +--3
+--5
   +--|
   +--6

putStrLn $ showTreeWith True False $ fromList [1..7]
4
+--2
|  +--1
|  +--3
+--6
   +--5
   +--7

putStrLn $ showTreeWith True False $ fromList [1..11]
8
+--4
|  +--2
|  |  +--1
|  |  +--3
|  +--6
|     +--5
|     +--7
+--10
   +--9
   +--11
`

**Comment by AlekJSmith** [comment]
@treeowl the draft function for the maximal element of the lexicographic ordered set is below. One caveat this builds for all the subsets of a given length not just those starting with a given minimum element. However the minimum element can be determined with Pascals triangle so I am holding off on that part until we meet can integrate this with your code.

Given a lexicographic ordered set A of size n, it has subsets of size k.  

The maximal element of the set of the highest value of the set and given the lexicographic ordering this value will be the last element of the set.

The pattern for the maximal element of the subsets of set A potentially can be constructed from a recursion of appending the tail of the previous result for example:

For set A of size 5 and subsets of length 1:
[1,2,3,4,5]

For set A of size 5 and subsets of length 2:
[2,3,4,5]++[3,4,5]++[4,5]++[5]

For set A of size 5 and subsets of length 3:
(
(
[3,4,5]++[4,5]++[5]
) ++
(
[4,5]++[5]
)
) ++ 
(
[5]
)

For set A of size 5 and subsets of length 4:
(
(
[4,5]++[5]
) ++ 
(
[5]
)
) ++ 
(
[5]
)

For set A of size 5 and subsets of length 5:
[5]


The complexity depends on tail, append and correct bracketing. With a deque the first two could be o(1) operations and would only need to be carried out k times, however the correct bracketing requires list nesting/concat which is o(n). Given the end result is planned to be turned into an array anyway this complexity should be ok.      


Version 1: List based

data Nested a = In a | Out [ Nested a ] deriving (Show, Eq)

fromNested (In a) = a 
fromNested (Out a) = a 

tail_append [] = [] ; tail_append x = [x] ++ ( tail_append (tail x) )

tails_nested (Out x) = Out (Prelude.map (\p -> Out p) (tail_append x))


max_elem n k = Prelude.map (\(In a) -> a) $ last $ Prelude.take k $ iterate (\p -> concat $ Prelude.map fromNested p) $ fromNested $ last $ Prelude.take k $ iterate tails_nested (Out [ In c | c <- [k..n] ])


#### Test

maximal_index n k = Prelude.map last $ Prelude.filter ((==k).length) $ Prelude.map toList $ toList $ powerSet $ fromList [1..n]

test1 n k = (max_elem n k) == (maximal_index n k)

test2 n = concat $ Prelude.map (\(x,y) -> Prelude.map (\g -> test1 x g) y) $ Prelude.zip [1..n]  [ [1..c] |  c <- [1..n] ]

any (==True) (test2 10)
